### ISSUE TITLE
[Android] Discrepancy between expo client and standalone in how keyboard affects view

### ISSUE THREAD

**Issue by seanpascoe** [expo/expo#5073]
## 🐛 Bug Report

yes this is similar and probably related to #2172, but that issue is from 2018 and I didn't start having this problem until recently. Though, I do not know which sdk or change is the culprit. (I upgrade to latest sdk within weeks of them coming out)

### Environment
```
  Expo CLI 3.0.4 environment info:
    System:
      OS: Linux 5.0 Ubuntu 19.04 (Disco Dingo)
      Shell: 5.0.3 - /bin/bash
    Binaries:
      Node: 10.16.0 - /usr/bin/node
      Yarn: 1.17.3 - /usr/bin/yarn
      npm: 6.9.0 - /usr/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    npmPackages:
      expo: ^33.0.0 => 33.0.7 
      react: 16.8.3 => 16.8.3 
      react-native: https://github.com/expo/react-native/archive/sdk-33.0.0.tar.gz => 0.59.8 
```
Target: Android

### Steps to Reproduce

1. Create app `expo init`
2. Add a `<TextInput/>` to app.js.
3. Add `"androidStatusBar": { "backgroundColor": "#000"},` to app.json.
4. Run app in expo client (android), click into TextInput and witness the view avoiding keyboard greatness. 
5. Run expo build:android. Install the apk. Witness the lack of avoiding keyboard greatness. 

### Expected Behavior
Ultimate expectation: expo client app behavior to be the same as the standalone app.
  
I have always used the androidStatusBar setting with opaque color in app.json. I toggle between a regular `<View/>` and `<KeyboardAvoidingView/>` depending on android/iOS, so my expectation is that this would work as it has for the past few years: iOS needs `<KeyboardAvoidingView behavior='padding' keyboardVerticalOffset={offset + statusBar}/>` to shrink the view and android works with either a simple `<View/>` or `<KeyboardAvoidingView/ >` (with behavior as null). 
It's hard to say what the expected behavior /should/ be. Inconsistencies between iOS and Android have been around since the beginning. The fact that settings in app.json (androidStatusBar, splash-- see closing comment in #2242, etc) effect the markup and ui behavior is a screwy thing in itself. 

### Actual Behavior
Expo Client: View shrinks when keyboard is active.
Standalone: View does not shrink when keyboard is active.

### Reproducible Demo

This issue cannot be reproduced via snack, as the problem lies in the discrepancy between expo app and standalone. Despite this, I hope the expo dev team can look at this one. It is very quick and easy to reproduce and it has been a total pain building android apk's over and over again.

**Comment by Slapbox** [comment]
Hmm... Maybe the cause of this suddenly appearing for me wasn't SDK 33, but me changing `androidStatusBar`... I forgot about that.

**Comment by Slapbox** [comment]
I forgot. This is definitely caused by changes to `androidStatusBar`. Whatever Expo magic there is in place that we end users don't understand, the magic broke here. It shocks me that that affects keyboard avoiding views by breaking them completely, rather than just being offset by the height of the Android Status Bar.

**Comment by cruzach** [comment]
Hey @seanpascoe - I just repro'd this on my end, can confirm the same behavior

One question- I tried it without setting `androidStatusBar` in `app.json`, and had the exact same behavior. Have you tried a plain react native app build to see if the same behavior happens there?

**Comment by seanpascoe** [comment]
@cruzach 
> I tried it without setting androidStatusBar in app.json, and had the exact same behavior.

You mean the view still shrunk when you removed `androidStatusBar` for expo client (testing on android phone)? Make sure you close the app and completely reload. I don't think app.json changes take effect unless you do. 
For me there is definitely a difference between `"androidStatusBar": { "backgroundColor": "#000"}` present vs. not in app.json (I think it has to be an opaque color).

**Comment by cruzach** [comment]
Oh. No, I meant for a standalone app I tried both with and without `"androidStatusBar": { "backgroundColor": "#000"}` set in app.json, and both resulted in no avoiding keyboard greatness.

In Expo Client-  with `"androidStatusBar": { "backgroundColor": "#000"}` set in app.json, I get the keyboard avoidance. without it, I don't

**Comment by seanpascoe** [comment]
Oh, ok yea, well that's the discrepancy: without `androidStatusBar`, the app behaves the same in expo and standalone... but with `androidStatusBar` the app behaves differently in expo vs standalone. To me, this signifies an issue with, as @Slapbox said, the 'Expo magic' happening in app.json and building the apk. I'm not sure what there is to test in a pure react native environment--you can't set `androidStatusBar` in app.json, right?

**Comment by Slapbox** [comment]
@cruzach when you say standalone, do you mean a non-Expo React Native project? I think @seanpascoe is using standalone to mean the APK file.

For me, the keyboard avoidance always works in development with `androidStatusBar` set, but does not work in an APK file. Unfortunately, setting production mode doesn't allow me to replicate that behavior, which means the only way to create the behavior is to do a full `expo publish` each time. 

To discover the cause of this, I eventually had to implement code in my app to allow me to choose avoidance modes on the fly. Needless to say, it wasn't easy to debug, as simple as the cause is.

**Comment by cruzach** [comment]
No, by standalone I mean an Expo-created app **not** running in the Expo Client app (so either an apk or ipa)

**Comment by seanpascoe** [comment]
@cruzach does this issue still `need more info`?

**Comment by Thomas-Wildeboer** [comment]
Any workarounds?

**Comment by b4stien** [comment]
@cruzach You added the "Client" label, but is it really an Expo client issue?

I was under the impression it was. a standalone app issue. Notably because, as of today, in a standalone app there is no way to have a working `KeyboardAvoidingView`.

**Comment by gerardo15** [comment]
I had the KeyboardAvoidingView working on my standalone app SKD32, i updated today to SDK34 and now its broken, i know that to make it work was the conbination of resizeMode:contain and androidStatusBar being set. but now. i have no clue how to fix it.

**Comment by cruzach** [comment]
@b4stien there's a discrepancy between the behavior of the two, so that's why I added the "Client" label. 

Also, as a workaround for having a working `KeyboardAvoidingView` -> just use a `KeyboardAvoidingView` rather than a regular `View` on Android standalones. I just tested this, and it works whether or not the `androidStatusBar` key is set in app.json

**Comment by gerardo15** [comment]
@cruzach I use KeyboardAvoidingView ands it's broken. thats not a workaround.

**Comment by Thomas-Wildeboer** [comment]
@gerardo15  @cruzach KeyboardAvoidingView is broken in my case as well, using the androidStatusBar key in app.json.

**Comment by cruzach** [comment]
On what SDK? I just built an APK on both SDK 32, and 34, tested on Android 9, with
```
"androidStatusBar": {
      "barStyle": "light-content",
      "backgroundColor": "#000000"
    }
```
set in app.json, and see the view scoot up above the keyboard

**Comment by Thomas-Wildeboer** [comment]
@cruzach SDK 33, I can update to 34 and see how that works

**Comment by b4stien** [comment]
Does `expo build:android` works for SDK 34 already? https://blog.expo.io/expo-sdk-34-is-now-available-4f7825239319 says no.

**Comment by baticpro** [comment]
Anyone solved this problem ? Maybe temporary decision ?

**Comment by b4stien** [comment]
I'm under the impression that the standalone built app on Android are in `adjustPan` which means, exactly as on iOS, that the window is not automatically resized and the keyboard slides "on top of" the activity window.

The Expo client on Android is in `adjustResize` mode, hence the difference.

We're trying to work around it by handling a standalone built app on Android exacty as an iOS app (with carefuly crafted `KeyboardAvoidingView`)...

By the way, it would be cool to document the interaction between `androidStatusBar` in `app.json` and the resize mode, there are some cryptic messages on the web, like: https://forums.expo.io/t/how-to-set-android-windowsoftinputmode-adjustresize-in-android-manifest-file/587/3.

1. Mode definitions : https://developer.android.com/guide/topics/manifest/activity-element.html#wsoft
2.

**Comment by JPig** [comment]
fwiw i tried removing the ```splash``` property from my ```app.json``` to see what happens and it solved the issue i was having in #5157 , perhaps it will help debug this issue also?

Mind you my issue isn't exactly "fixed", as i now don't have a splash screen haha.

**Comment by Thomas-Wildeboer** [comment]
If you don't define android status bar in app.json the KeyboardAvoidingView is broken in SDK34 after building.

**Comment by gerardo15** [comment]
@Thomas-Wildeboer I have defined android status bar in app.json and KeyboardAvoidingView  is broken for me in SDK34

**Comment by EtienneBourque** [comment]
I also had this problem on SDK 34 with the android status bar defined in app.json and the splash in `contain` mode.

> I'm under the impression that the standalone built app on Android are in `adjustPan` which means, exactly as on iOS, that the window is not automatically resized and the keyboard slides "on top of" the activity window.

Now, reading this reply, and despite knowing that usually the consensus was that on Android, the KeyboardAvoidingView didn't want the `behavior="padding"` attribute, I added it and now the keyboard is pushing back my screen correctly in the APK build. However, the screen is pushed back too much running in the Expo client.

Without the `behavior` attribute, the screen is not pushed back in the APK build and is pushed correctly in the Expo client.

I did another test. Using a regular view but putting the Android splash screen in `native` mode, the keyboard pushed back the screen correctly both in the APK build and the Expo client. Now though, my splash screen is way too big in the APK build but correct in the Expo client.

**Comment by rabihiawaludin** [comment]
@EtienneBourque have you solved it?

**Comment by rabihiawaludin** [comment]
any update?

**Comment by EtienneBourque** [comment]
> @EtienneBourque have you solved it? i have this issue too

Well, for now, putting the Android splash screen in native mode (see the Expo docs app.json page) solves the keyboard not pushing back the view. No KeyboardAvoidingView needed. 

The downside is that my splash image is not resized anymore (normal for the native mode) so it's showing way too big (only in APK). I could include smaller versions in my assets.

I'm not sure if this new behavior is supposed to be normal now. But for sure, it's not consistent between the Expo client and the APK build.

**Comment by swayok** [comment]
I had KAV working correctly in 33 SDK for some time but it became broken later and it is broken in 34 SDK. Native mode acts weirdly (very large image) even when I provide all image versions from x1 to x4.
I manged to get several workarounds but nothing works normally right now =( Splash screens are totally broken in Android.

**Comment by Slapbox** [comment]
@cruzach this is labeled `Needs More Info`, can you tell us what additional info though? It seems like nothing will get done until that status changes, and I think users have given all the info they can. 

The only info we need at this stage, it seems to me, is what change to Expo, or what interplay between Expo and React Native, was the genesis of the issue. No one with the issue can provide that info though. I know I would happily help if I had the faintest clue, but I haven't.

At current, and I think I speak for a number of people, my Expo app has become a development project rather than a production ready app. It's is basically unshippable. KAV is pretty frustrating to deal with at times on 1 OS, let alone 2 OS's; then add in that various screens might need different `behavior`s. Then we get differences in `Dev` and `Prod`, but you can't reproduce this in `Prod` mode, you can only do a test in a built project. (Side grumble, Prod mode requires me to nuke the Expo Client App to escape it if I ever turn it on)

I don't mean to sound ungrateful at all, I think the Expo team is amazing and really helpful and friendly to developers and their needs - it just seems like the seriousness of this issue isn't being recognized. This issue breaks Expo in a major way. I really love Expo, and you guys for your hard work. Expo's otherwise been an awesome experience.

**Comment by cruzach** [comment]
@Slapbox I removed the `needs more info` label awhile ago, are you sure you're still seeing it? You're definitely right in that it seems the only remaining info to be found is the root cause, and the corresponding fix!

I appreciate you voicing your thoughts here, and we recognize how serious this issue is (anything that results in a difference in behavior between the Expo Client and standalone apps is severe). Rest assured we're looking into it and hopefully will have more news for you soon

**Comment by swayok** [comment]
One more problem that I'm experiencing only in APK version is that splash screen in cover mode shifts a bit towards bottom after app is loaded. I'm using `SplashScreen.preventAutoHide();` and it is clearly visible. Also I've tested how it works without `SplashScreen.preventAutoHide();` and it also shifts splash screen just before dissapearing. Visually difference seems like status bar height but on some devices it is less then status bar. In Expo Client this is not happening. This issue was in SDK 32 but dissapeared in early versions of SDK 33 and (if I correctly remember) reappeared some time after expo-cli updated to version 3.

I'm using next config:
```
"splash": {
     "backgroundColor": "#a3d8ff"
},
"androidStatusBar": {
    "barStyle": "dark-content",
    "backgroundColor": "#a3d8ff"
},
"android": {
    "splash": {
        "backgroundColor": "#a3d8ff",
        "resizeMode": "cover",
        "mdpi": "./assets/images/splash_screen/simple/splash.png",
        "hdpi": "./assets/images/splash_screen/simple/splash@1.5x.png",
        "xhdpi": "./assets/images/splash_screen/simple/splash@2x.png",
        "xxhdpi": "./assets/images/splash_screen/simple/splash@3x.png",
        "xxxhdpi": "./assets/images/splash_screen/simple/splash@4x.png"
    }
},
```
Images:
- original - 720x1280 
- 1.5x - 1080x1920
- 2x - 1440x2560
and so on.

Thank you wor your work! Expo is great in many ways =)

**Comment by swayok** [comment]
Just noticed one more thing: KAV works correctly with Modals.

**Comment by mxhold** [comment]
We ran into this and, like @JPig, found that removing the splash screen entirely for Android gets you back to the right behavior (keyboard causes view to shift up without the need for a KAV on Android). It seemed preferable to do this workaround and sacrifice having a splash screen in order for the keyboard interactions to work correctly, though having both work would of course be preferable.

We did this by moving our previously top-level `splash` configuration from `app.json` to just be nested under `ios` to keep the splash screen working on iOS. Then when the app launches on Android, there's just an empty white splash screen but the keyboard causes the view to shift up as it does when running via the Expo client.

Oddly, right after we released this, the splash screen actually continued to show for some time (though strangely it seemed to have switched to use a `resizeMode` of `contain` instead of what we previously specified, `cover`) and the keyboard continued to act incorrectly. I thought this might be due to the way over-the-air updates work but even uninstalling and reinstalling the standalone app right after releasing seemed not to get the new version. However, when I opened the app the next day, the keyboard seemed to be fixed and the app was no longer showing the splash screen. Not sure what that's about or if that's unexpected, maybe there's something about the release process I still don't understand.

**Comment by Slapbox** [comment]
@cruzach thanks so much for the continued help!

**Comment by powerpovar** [comment]
Have the same issue in SDK33.
Any updates?

**Comment by dcefram** [comment]
It's a shame that the latest Expo client does not allow running SDK 32, which was the last time the `KeyboardAvoidingView` used to work. I'm now forced to build the app and then load it in the phone rather than just running in through Expo client. 

This would've been a non-issue if this bug was fixed before removing the support of the last known working SDK version. It's frustrating, but I could work around with it by compiling an APK... It would be great if there's some .apk file of older expo clients out there, I can't seem to find them :|

**Comment by giacomocerquone** [comment]
@dcefram you can find an old version of the expo client here: https://blog.expo.io/expo-sdk-35-is-now-available-beee0dfafbf4

```
If you still need to test a project running SDK 31 or 32, you can download this Android client version and this iOS client version.
```

android: https://d1ahtucjixef4r.cloudfront.net/Exponent-2.12.2.apk
ios: https://dpq5q02fu5f55.cloudfront.net/Exponent-2.12.3.tar.gz

**Comment by terrencejames** [comment]
I can confirm that @mxhold 's solution about removing the splash configuration in the app.json fixes the issue that I have, which is similar except that it's with views without the keyboard as well, so perhaps this issue goes beyond keyboard views

**Comment by bbarthec** [comment]
I'm closing this issue as it's duplicate of #2172 and both issues are addressed by #6506.
